### PR TITLE
Add support for asynchronous elements

### PR DESCRIPTION
**Pull request by tjjfvi**

### CODE REVIEW

**Comment by tjjfvi** [comment]
Closes #11

**Comment by tjjfvi** [comment]
Fixes #11

File: packages/core/src/Hierarchy.ts [hunk 1]
```diff
@@ -33,7 +33,8 @@ export const Hierarchy = {
     || NameHierarchy.isNameHierarchy(value)
     || ValueHierarchy.isValueHierarchy(value)
     || CallHierarchy.isCallHierarchy(value),
-  from: (value: unknown, raw = false): Hierarchy => {
+  from: async (value: unknown, raw = false): Promise<Hierarchy> => {
+    value = await value
```
Comment (vjjft) [inline packages/core/src/Hierarchy.ts:40]: :/

File: packages/core/src/Element.ts [hunk 2]
```diff
@@ -61,69 +70,82 @@ export const Element = {
...
+    if(ObjMap.isObjMap(value))
+      return Object.values(value)
+    if(Product.isProduct(value))
       return [element]
     throw new Error("Invalid Element.val type")
   },
-  toArrayDeep: <T extends Product>(element: Elementish<T>): DeepArray<T> | T => {
+
+  toArrayDeep: async <T extends Product>(element: Elementish<T>): Promise<DeepArray<T> | T> => {
     element = Element.create(element)
```
Comment (vjjft) [inline packages/core/src/Element.ts:103]: :/

File: packages/core/src/Element.ts [hunk 3]
```diff
@@ -61,69 +70,82 @@ export const Element = {
...
+    const elementPromise = Element.createPromise(elementish, hierarchy)
+    return {
+      type: "Element",
+      value: elementPromise.then(el => el.value),
+      hierarchy: elementPromise.then(el => el.hierarchy),
+    }
+  },
+
+  toArray: async <T extends Product>(element: Elementish<T>): Promise<Array<Element<T>>> => {
     element = Element.create(element)
```
Comment (vjjft) [inline packages/core/src/Element.ts:91]: :/

File: packages/core/src/Element.ts [hunk 4]
```diff
@@ -30,16 +31,24 @@ interface DeepArray<T> extends ReadonlyArray<DeepArray<T> | T> {}
...
-  create: <T extends Product>(elementish: Elementish<T>, hierarchy?: Hierarchy): Element<T> => {
+
+  createPromise: async <T extends Product>(
+    elementish: Promisish<Elementish<T>>,
+    hierarchy?: Promisish<Hierarchy | undefined>,
+  ): Promise<Element<T>> => {
     let value: T | Arrayish<Element<T>>
+
+    elementish = await elementish
+    hierarchy = await hierarchy
```
Comment (vjjft) [inline packages/core/src/Element.ts:59]: :/

**Review by vjjft — CHANGES_REQUESTED** [review]

File: packages/client/src/HierarchyView.tsx [hunk 5]
```diff
@@ -245,7 +248,7 @@ function _hierarchyToTree(hierarchy: Hierarchy): Tree{
       { text: ["("] },
       { children: operators, joiner: "∘", state: { open: false }, forceOpenable: true },
       { text: [")"] },
-      ...wrapLinkedProducts(ArrayHierarchy.from(operands), [
+      ...wrapLinkedProducts(hierarchy.operands.flatMap(x => x.linkedProducts ?? []), [
```
Comment (vjjft) [inline packages/client/src/HierarchyView.tsx]: I suggest extracting to Hierarchy.flattenLinkedProducts

File: packages/client/src/HierarchyView.tsx [hunk 6]
```diff
@@ -203,7 +206,7 @@ function _hierarchyToTree(hierarchy: Hierarchy): Tree{
     if(!hierarchy.composable)
       return [
         ...hierarchyToTree(hierarchy.operator),
-        ...wrapLinkedProducts(ArrayHierarchy.from(hierarchy.operands), [
+        ...wrapLinkedProducts(hierarchy.operands.flatMap(x => x.linkedProducts ?? []), [
```
Comment (vjjft) [inline packages/client/src/HierarchyView.tsx]: I suggest extracting to Hierarchy.flattenLinkedProducts

File: packages/client/src/HierarchyView.tsx [hunk 7]
```diff
@@ -148,21 +150,22 @@ type TreeTextPartProps = {
   className?: string,
 }
 
-function wrapLinkedProducts(hierarchy: Hierarchy, tree: Tree): Tree{
-  if(!hierarchy.linkedProducts || !hierarchy.linkedProducts.length)
+function wrapLinkedProducts(linkedProducts: readonly Hash<Product>[] | undefined, tree: Tree): Tree{
```
Comment (vjjft) [inline packages/client/src/HierarchyView.tsx]: I suggest extracting type for linkedProducts

File: packages/core/src/Component.ts [hunk 8]
```diff
@@ -5,20 +5,21 @@ import { contextStack } from "./ContextStack"
...
 import { NameHierarchy } from "./NameHierarchy"
+import { Promisish } from "./Promisish"
 import { Thing, StripRealm } from "./Thing"
 
 export interface Component<I extends any[], T extends Thing> {
   readonly type: "Component",
   readonly func: (...input: I) => T,
   readonly name: string,
-  readonly hierarchy?: Hierarchy,
+  readonly hierarchy?: Promisish<Hierarchy | undefined>,
```
Comment (vjjft) [inline packages/core/src/Component.ts]: I suggest extracting type

File: packages/core/src/Element.ts [hunk 9]
```diff
@@ -61,69 +70,82 @@ export const Element = {
 
     return {
       type: "Element",
-      value,
-      hierarchy,
+      value: Promise.resolve(value),
+      hierarchy: Promise.resolve(hierarchy),
```
Comment (vjjft) [inline packages/core/src/Element.ts]: from discussion, @tjjfvi suggested using `Promisish` for type as done elsewhere and removing `Promise.resolve`

File: packages/core/src/Element.ts [hunk 10]
```diff
@@ -61,69 +70,82 @@ export const Element = {
...
+    if(ObjMap.isObjMap(value))
+      return Element.create(Object.fromEntries(Object.entries(value).map(([k, v]) => [k, Element.map(v, fn)])))
+
+    if(Product.isProduct(value))
+      return Element.create(fn(value))
 
     throw new Error("Invalid Element.val type")
   },
+
+  map: <T extends Product, U extends Product>(element: Element<T>, fn: (value: T) => Elementish<U>): Element<U> =>
```
Comment (vjjft) [inline packages/core/src/Element.ts]: consider moving above `mapPromise`

File: packages/core/src/Operation.ts [hunk 11]
```diff
@@ -22,32 +23,22 @@ export interface Operation<I extends Product, O extends Product> {
...
+  readonly hierarchy?: Promisish<Hierarchy | undefined>,
   readonly overrideHierarchy?: boolean,
   readonly showOutputInHierarchy?: boolean,
 }
 
 export const Operation = {
   create: <I extends Product, O extends Product>(
     name: string,
-    func: (arg: Element<I>) => Elementish<O>,
+    func: (arg: Element<I>) => Promisish<Elementish<O>>,
```
Comment (vjjft) [inline packages/core/src/Operation.ts:34]: consider extracting type to `Element.ts`

**Review by vjjft — DISMISSED** [review]

File: packages/core/src/Operation.ts (hunk 11 above)
Comment (tjjfvi) [reply packages/core/src/Operation.ts:34]: `Elementish` can't have `Promisish` baked in, as it was causing the inferred type of `Element` to be too large. I feel that `PromisishElementish` adds little, and other names would be confusing (e.g. `AsyncElementish` suggests that `Elementish` is fully synchronous), so I'm going to leave this in. It's only used in 3 places.

**Pull request merged**